Add hook for modifying OutputOptions

Comandeer opened the pull request:
<!--
  ⚡️ katchow! We ❤️ Pull Requests!

  If you remove or skip this template, you'll make the 🐼 sad and the mighty god
  of Github will appear and pile-drive the close button from a great height
  while making animal noises.

  Pull Request Requirements:
  * Please include tests to illustrate the problem this PR resolves.
  * Please lint your changes by running `npm run lint` before creating a PR.
  * Please update the documentation in `/docs` where necessary

  Please place an x (no spaces - [x]) in all [ ] that apply.
-->

This PR contains:
- [ ] bugfix
- [x] feature
- [ ] refactor
- [x] documentation
- [ ] other

Are tests included?
- [x] yes (*bugfixes and features will not be merged without tests*)
- [ ] no

Breaking Changes?
- [ ] yes (*breaking changes will not be merged unless absolutely necessary*)
- [x] no

List any relevant issue numbers:

#2735

### Description

This PR adds `outputOptions` hook for plugins, that allows overriding of `OutputOptions` passed to `bundle.generate`. I've based the code on the existing `options` hook. The hook is run inside `normalizeOutputOptions`, as it seemed to be the place, where the final form of output options is created.

lukastaegert commented:
Hi @Comandeer, this looks like a valuable addition. However I would propose to make this a "proper" hook using a plugin context which enables the hook to e.g. display warnings using Rollup's internal mechanism. The original `options` hook was a rather bad example to copy from because this hook cannot use the proper context as many fundamental things including plugins are not known before this hook has run.

To do that, you use the `pluginDriver` which is attached to the graph. As I think it makes sense to keep the hook synchronous, we would need a new `hookReduceArg0Sync` to handle this.

lukastaegert commented on docs/05-plugins.md:
```diff
@@ -147,7 +147,13 @@ Defines a custom loader. Returning `null` defers to other `load` functions (and
+Type: `(outputOptions: OutputOptions) => OutputOptions | null`<br>
+Kind: `sync, sequential`
+
+Reads and replaces or manipulates the output options object passed to `rollup.rollup`. Returning `null` does not replace anything. This hook does not have access to most [plugin context](guide/en#plugin-context) utility functions as it is run before rollup is fully configured.
```
Thanks for taking care to update the documentation! Some comments:
- output options are not passed to `rollup.rollup` but to `bundle.generate`
- with my suggestion, the limitations can be removed

lukastaegert commented on src/rollup/index.ts:
```diff
@@ -66,6 +66,13 @@ function applyOptionHook(inputOptions: InputOptions, plugin: Plugin) {
+
+	return outputOptions;
+}
+
```
This can be handled by the pluginDriver

lukastaegert commented on src/rollup/index.ts:
```diff
@@ -427,7 +434,15 @@ function normalizeOutputOptions(
+		: plugins
+		? [plugins]
+		: [];
+	const outputOptions = inputOptions.plugins.reduce(applyOutputOptionHook, mergedOutputOptions);
```
This can be handled by the pluginDriver. To get access, you probably need to pass it to this function.

lukastaegert commented on src/rollup/types.d.ts:
```diff
@@ -230,6 +230,10 @@ export interface Plugin {
 	) => void | Promise<void>;
 	options?: (this: MinimalPluginContext, options: InputOptions) => InputOptions | void | null;
+	outputOptions?: (
+		this: MinimalPluginContext,
```
This can be the proper `PluginContext`

lukastaegert commented on test/hooks/index.js:
```diff
@@ -29,6 +29,35 @@ describe('hooks', () => {
+					{
+						renderChunk(code, chunk, options) {
+							assert.strictEqual(options.banner, 'new banner');
+							assert.strictEqual(options.format, 'cjs');
```
It would be nice to verify that not only the options are updated, but the code was actually rendered using the updated options by inspecting the generated code.

For the latter, I think it would be best to inspect the output of `bundle.generate`

Comandeer commented:
I've implemented `pluginDriver.hookReduceArg0Sync` method and used it inside `normalizeOutputOptions` to run `outputOptions` hook. I've also updated docs and tests according to your suggestion.

lukastaegert commented on src/rollup/index.ts:
```diff
@@ -427,7 +430,17 @@ function normalizeOutputOptions(
+	const outputOptionsReducer = (outputOptions: OutputOptions, result: OutputOptions) => {
+		if (!result) return outputOptions;
+
+		return {...outputOptions, ...result};
```
The rest looks really good, this is the only thing giving me pause. I am not sure we should be merging the output options; at least when compared to the options hook that only replaces options. Though I would mostly argue for consistency here.

Also, this makes it impossible to remove options (though of course you could set them to `undefined`), which could be confusing for some options where `undefined` and `false` behave differently (e.g. `output.interop`).
And lastly, this is only a shallow merge which may or may not be what people expect for object valued options (there are only few, e.g. `output.amd`).

Would you be ok with a change or do you have strong arguments for merging?

Comandeer replied:
No, I don't have any strong arguments. TBH I also had some doubts about it, yet it seemed like a more elegant solution to me. However you made good point about removing options. I simplified the code to just replace options.
